documentation/design-docs/diagnostics-client-library.md

```diff
@@ -0,0 +1,296 @@
+    }
+}
+```
+
+### DumpCommandHandler
+This is a CommandHandler class for sending Dump commands across the diagnostics IPC channel.
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    public class DumpCommandHandler
+    {
+        public DumpCommandHandler(int processId);
+        public int GenerateCoreDump();
+    }
+}
+```
+
+
+### ProfilerCommandHandler
+This is a CommandHandler class for sending Profiler commands across the diagnostics IPC channel.
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    public class ProfilerCommandHandler
+    {
+        public ProfilerCommandHandler(int processId);
+        public int AttachProfiler(uint attachTimeout, Guid profilerGuid, string profilerPath, byte[] additionalData);
+    }
+}
+```
+
+
+### Exceptions that can be thrown 
+```cs
+namespace Microsoft.Diagnostics.Client.Exceptions
+{
+    public class UnknownCommandException : Exception {}
+    public class UnknownMagicException : Exception {}
+    public class EventPipeInvalidArgumentException : Exception {}
+    public class DumpInvalidArgumentException : Exception {}
+    public class ProfilerInvalidArgumentException : Exception {}
+}
+```
+
+## Micorosft.Diagnostics.Client.EventPipe namespace
+
+### EventPipeCommandSet
+An enum for all EventPipe commands.
+``` cs
+namespace Micorosft.Diagnostics.Client.EventPipe
+{
+    public enum EventPipeCommandSet
+    {
+        StopTracing,
+        CollectTracing,
+        CollectTracing2
+    }
+}
+```
+
+### EventPipeSerializationFormat
+An enum for all EventPipe serialization format.
+```cs
+namespace Microsoft.Diagnostics.Client.EventPipe
+{
+    public enum EventPipeSerializationFormat
+    {
+        NetPerf,
+        NetTrace
+    }
+}
+```
+
+### EventPipeProvider
+A class that describes an EventPipe provider.
+```cs
+namespace Microsoft.Diagnostics.Client.EventPipe
+{
+    public class EventPipeProvider
+    {
+        public EventPipeProvider(
+            string name,
+            ulong keywords = ulong.MaxValue,
+            EventLevel eventLevel = EventLevel.Verbose,
+            string filterData = null)
+
+        public ulong Keywords { get; }
+
+        public EventLevel EventLevel { get; }
+
+        public string Name { get; }
+
+        public string FilterData { get; }
+
+        public string ToDisplayString();
+
+        public override string ToString();
+        
+        public override bool Equals(object obj);
+
+        public override int GetHashCode();
+
+        public static bool operator ==(Provider left, Provider right);
+
+        public static bool operator !=(Provider left, Provider right);
+    }
+}
+```
+## Sample Code:
+
+
+### 1. Attaching to a process and dumping out its event data 
+```cs
+
+public static void Main(String[] args)
+{
+    int processId = Int32.Parse(args[0]);
+    
+    // Create an EventPipe command handler for the given process ID
+    EventPipeCommandHandler handler = new EventPipeCommandHandler(processId);
```

Comment:
Can I create an EventPipe command handler for a process on a different machine? Or a process in a docker container?

Reply:
Not at the moment since this isn't a supported scenario at the moment. We're in the process of figuring it out.